### PR TITLE
Tests/functional

### PR DESCRIPTION
**Pull request by DenisMishin**
CLI based tests and infra

### CODE REVIEW

File: infra/infra.go [hunk 1]
```diff
@@ -126,11 +129,16 @@ type NodePool interface {
...
 // Node defines an interface to a remote node
 type Node interface {
-	// Addr returns the address of the node
+	// Addr returns public address of the node
 	Addr() string
+	// PrivateAddr returns the private address of the node
+	PrivateAddr() string
 	// Connect connects to this node and returns a new session object
 	// that can be used to execute remote commands
 	Connect() (*ssh.Session, error)
```
Comment (a-palchikov) [inline infra/infra.go]: We don't need `Connect` as part of `Node` anymore - one can always create a `Session` from a `Client`. Besides, the implementation is identical and hence duplicate in each provisioner.

I think internal unit-tests are out-of-date w.r.t changes to this interface.

File: lib/ssh/commands.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+// Cmd defines command to execute on remote host
+type Cmd struct {
+	// Command is what is passed to remote shell
+	Command string
+	// Env are any environment variables to be set
+	// Note that environment vars must be explicitly enabled in /etc/ssh/sshd.conf `AcceptEnv` directive
+	Env map[string]string
+}
+
+// RunCommands execute commands sequentially
```
Comment (a-palchikov) [inline lib/ssh/commands.go]: s/execute/executes as the function name is treated as just a name

File: lib/ssh/run_parse.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	if exit != 0 {
+		return trace.Errorf("%s returned %d", cmd, exit)
+	}
+
+	return nil
+}
+
+// RunAndParse runs remote SSH command with environment variables set by `env`
+// exitStatus is -1 if undefined
+func RunAndParse(ctx context.Context, logFn LogFnType, client *ssh.Client, cmd string, env map[string]string, parse OutputParseFn) (out interface{}, exitStatus int, err error) {
```
Comment (a-palchikov) [inline lib/ssh/run_parse.go]: The function expects a `ssh.Session`, and the logging detail can be setup prior to the call:
```go
session, err := client.NewSession()
if err != nil {
	return trace.Wrap(err, "failed to create a session")
}
entry := log.WithFields(log.Fields{"client": client.RemoteAddr()})
exitStatus, err := RunAndParse(ctx, session, cmd, env, parseFn, entry)

func RunAndParse(ctx context.Context, session *ssh.Session, cmd string, env map[string]string, parse OutputParseFn, entry *log.Entry) (out interface{}, exitStatus int, err error) {
	// ...
	entry.Debugf("stderr: %s", out)
}
```
This way you don't need to spill too much detail into the function itself.

And the function itself feels too heavy.

File: lib/ssh/run_parse.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+		for {
+			line, err := r.ReadString('\n')
+			if err != nil {
+				return
+			}
+			logFn("[%v] %s [stderr] %s", client.RemoteAddr(), cmd, line)
+		}
+	}()
+
+	outCh := make(chan interface{}, 2)
```
Comment (a-palchikov) [inline lib/ssh/run_parse.go]:
```go
var outCh chan resp

type resp struct {
	out []byte
	err error
}
```
why the buffer of 2? There's just a single producer

File: lib/ssh/run_parse.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	go func() {
+		logFn("[%v] (starting) %s", client.RemoteAddr(), cmd)
+		runCh <- session.Run(fmt.Sprintf("%s %s", strings.Join(envStrings, " "), cmd))
+	}()
+
+	select {
+	case <-ctx.Done():
+		return nil, exitStatusUndefined,
+			trace.Errorf("[%v] %s timed out, sending SIGTERM: %v",
+				client.RemoteAddr(), cmd, session.Signal(ssh.SIGTERM))
```
Comment (a-palchikov) [inline lib/ssh/run_parse.go]: can you extract the `Signal` call (so it's not part of `Errorf` call)?

File: lib/ssh/ssh_test.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		client,
+		"sleep 100",
+		nil,
+		ParseDiscard)
+	assert.Error(t, err)
+	assert.Equal(t, exitStatusUndefined, exit)
+}
+
+func testExitErr(t *testing.T, client *ssh.Client) {
+	_, exit, err := RunAndParse(context.Background(), t.Logf, client, "ls /nosuchdir", nil, ParseDiscard)
```
Comment (a-palchikov) [inline lib/ssh/ssh_test.go]: better use `false` as an always failing command

File: suite/install.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+		go func(n gravity.Gravity) {
+			// TODO: how to properly define node role ?
+			errs <- n.Join(ctx, gravity.JoinCmd{
+				PeerAddr: master.Node().PrivateAddr(),
+				Token:    token,
+				Role:     defaultRole})
+		}(node)
+	}
+
+	for range nodes {
```
Comment (a-palchikov) [inline suite/install.go]: Handle nodes never returning from install/join call with a timeout?

File: infra/gravity/gravity_test.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+	sshutils "github.com/gravitational/robotest/lib/ssh"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGravityOutput(t *testing.T) {
+	for _, s := range []struct {
+		label string
+		fn    sshutils.OutputParseFn
```
Comment (DenisMishin) [inline infra/gravity/gravity_test.go]: as discussed will be changed after implemented in gravity repo in next few days.

File: infra/gravity/log.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+func init() {
+	log.SetFlags(0)
+}
+
+func Logf(t *testing.T, prefix string) sshutils.LogFnType {
+	return func(format string, args ...interface{}) {
+		t.Logf(format, args...)
+		log.Printf(fmt.Sprintf("[%s] %s", prefix, format), args...)
+	}
+}
```
Comment (DenisMishin) [inline infra/gravity/log.go]: as discussed, will stay as is: in non-verbose mode, will just be equal to t.Logf and thus will log line numbers correctly.

File: infra/gravity/provision.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+	dockerDevice string
+	user         string
+	os           string
+	installDir   string
+	installerUrl string
+	tf           terraform.Config
+	env          map[string]string
+}
+
+// makeDynamicParams takes base config, validates it and returns TerraformConfig
```
Comment (DenisMishin) [inline infra/gravity/provision.go]: it's not related to just terraform, so naming is right.

File: infra/terraform/terraform.go [hunk 11]
```diff
@@ -250,6 +249,7 @@ func (r *terraform) boot() (output string, err error) {
 	out, err := r.command([]string{
 		"apply",
 		"-var", fmt.Sprintf("nodes=%d", r.NumNodes),
+		"-var", fmt.Sprintf("os=%s", r.OS),
```
Comment (DenisMishin) [inline infra/terraform/terraform.go]: as discussed, TF destroy is not very reliable method of tearing down, and it'll be external service which would ensure cleanup by timeout. anyway right now destory is a no-op, and it'll be addressed in next few days.

File: infra/infra.go (hunk 1 above)
Comment (DenisMishin) [inline infra/infra.go]: for legacy code compatibility

File: suite/install.go (hunk 7 above)
Comment (DenisMishin) [inline suite/install.go]: handled by ctx

File: infra/gravity/provision.go (hunk 10 above)
Comment (a-palchikov) [inline infra/gravity/provision.go]: I meant that that the original comment states it returns a TerraformConfig while it does not:
`// makeDynamicParams takes base config, validates it and returns TerraformConfig`

File: Makefile [hunk 12]
```diff
@@ -1,82 +1,62 @@
...
+docker-images:
+	cd docker && $(MAKE) publish
 
 .PHONY: publish-binary-into-s3
 publish-binary-into-s3:
-ifeq (, $(shell which aws))
-	$(error "No aws command in $(PATH)")
-endif
-	aws $(S3_OPTS) s3 cp ./build/robotest $(BUILD_BUCKET_URL)/robotest
+	ifeq (, $(shell which aws))
```
Comment (UnderGreen) [inline Makefile]: It's not a nitpick. Without normal indentation `ifeq` will not work. I know, because I've found this issue several weeks ago.

**Pull request merged**